app/controllers/users/sessions_controller.rb

```diff
@@ -13,19 +16,26 @@ def create
   end
 
   def sign_in_with_token
-    user = User.find_by(login_token: params[:login_token])
-
-    if user.present? && Time.zone.now < user.login_token_valid_until
-      user.update!(login_token: nil, login_token_valid_until: 1.year.ago)
-      sign_in(user, scope: :user)
-      redirect_to helpers.profile_dashboard_url(user)
-    else
-      flash[:alert] = "There was an error while logging you in. Please enter your email again."
-      redirect_to new_user_session_path
-    end
+    @user.update!(login_token: nil, login_token_valid_until: 1.year.ago)
+    sign_in(@user, scope: :user)
+    redirect_to_dashboard
   end
 
   def redirect_from_magic_link
     @login_token = params[:login_token] if params[:login_token].present?
   end
+
+private
+
+  def redirect_to_dashboard
+    redirect_to helpers.profile_dashboard_url(current_user) if current_user.present?
+  end
+
+  def ensure_login_token_valid
+    @user = User.find_by(login_token: params[:login_token])
+    unless @user.present? && Time.zone.now < @user.login_token_valid_until
```

Comment:
Small stylistic for better readability - adding an empty line improves the speed of reading the code when there is a block. A case of oranges over apples. You could also put Time.zone.now < @user.login_token_valid_until in a variable for better readability or even put it to user class. An opinion - not a hard rule.

Reply:
I prefer to use ! @user.present?  in this case rather than  unless @user.present? . When we have a compound predicate with unless I can never quite remember if the unless applies to (@user.present? && Time.zone.now < @user.login_token_valid_until) or just to @user.present?.  I find ! @user.present? just clearer

Reply:
Or maybe `if @user.blank? or Time.zone.now > @user.login_token_valid_until`, hehe. So many ways to write a conditional in Ruby.

I would second putting the login token time check outside of this line. My brain gets dizzy with more than two logical complexities at the same time - one being `unless` and the other joint statement without a clear name.

Reply:
if @user.blank? works for me too and is better than  ! @user.present?

Reply:
This is one of the only lines I didn't change 😂 I'll smarten it up 👍